### ISSUE TITLE
The `Problem Formulation.ipynb` notebook has weird Policy data retention issues, potentially corrupting results 

### ISSUE THREAD

**Issue by EwoutH** [quaquel/epa1361_open#5]
@anneheijbroek and I have been breaking my neck over this issue for the better part of a day. There seems something data retentions quicks in the `dike_model` object, and then especially it's policies (`dike_model.policy`), in the [Problem Formulations.ipynb](https://github.com/quaquel/epa1361_open/blob/cbc0e34a78cd73898c1a11a6a483ed84ef4cc114/final%20assignment/Problem%20Formulations.ipynb) notebook.

In cell 8 the `dike_model` is first run, with 4 (random) generated policies. Before this run, the `dike_model.policy` attribute is empty, after this run, it's filled with a `Policy()` object which contains a dictionary.

Now the problem is, it seems that those 4 generated policies that are now in `dike_model.policy` are taken over to the second run with the custom defined policies! And that breaks all kind of things.

The obvious solution is removing the first run with the generated policies, but then the model crashes in the `dike_model_function.py`, where the `progressive_height_and_costs` function expects for each dike to have a dike increase key-value pair ready. It crashes on https://github.com/quaquel/epa1361_open/blob/cbc0e34a78cd73898c1a11a6a483ed84ef4cc114/final%20assignment/dike_model_function.py#L94

So here's the issue: The notebook in the current form seems to take the old, generated `dike_model.policy` dictionary keys `f'DikeIncrease {s}'` for every dike that wasn't defined by hand. So the results are all over the place.

@quaquel am I doing something totally wrong or isn't this expected behavior?

### How to reproduce
In the [Problem Formulations.ipynb](https://github.com/quaquel/epa1361_open/blob/cbc0e34a78cd73898c1a11a6a483ed84ef4cc114/final%20assignment/Problem%20Formulations.ipynb) notebook:

- Readout `dike_model.policy.data` after first and second run, and notice how the values are (largely) the same, or
- Remove `policies=4` in the first run (cell 8) and note the second run crashes with a KeyError
- Remove cell 8 and 9 all together and note the second run crashes with a KeyError

Note, if the `MultiprocessingEvaluator` keeps running endlessly, replace with `SequentialEvaluator` to see it crash.

**Comment by quaquel** [comment]
Not sure I fully understand the issue without trying to reproduce it. However, you must make sure that custom policies have values for all levers. This value can of course be 0. 

The model.policy attribute is generic (part of the inner workings of the workbench) and is updated when you are evaluating a new policy. See `experiment_runner.py` in the ema_workbench and `AbstractModel.model_init`.

**Comment by EwoutH** [comment]
> However, you must make sure that custom policies have values for all levers.

Right, that really is counter intuitive in the Problem Formulations notebook, since in the second example with custom policies only a few specific levers are set, which implies the other values are either random or a default value and thus don’t have to be explicitly set. However, it looks like the values are copied over from the previous run in cell 8.

**Comment by quaquel** [comment]
I figured out what is causing this and it is indeed not desired behavior. In short, in `DikeNetwork.__call__`, the code sets G to self.G (see line 120 in dike_model_funciton.py). However, dike raising options are set as attributes on the node in the network (see line 157 in dike_model_function.py). This attribute persists for a next policy unless explicitly overwritten. The easiest fix is to replace `G = self.G` with `G = copy.deepcopy(self.G)`. This will cause the notebook to throw an error in cell 9 because some levers are missing.

Still have to check the knock on consequences of this in other parts of the example code and fix those before pushing this fix.

**Comment by EwoutH** [comment]
Thanks for getting back, and implementing the fix above.

Could we prevent manually having to define all Policy levers by assigning a default value to the RfR, EWS and dike levers (in problem_formulation.py)? If so, how do we parse this when defining policies? Are there build in functions to use the default levers for in a Policy and modify only a few lever values?

https://github.com/quaquel/epa1361_open/blob/91d2b63942c576aaf617bddedd55286fb3c72a04/final%20assignment/problem_formulation.py#L56-L65

https://github.com/quaquel/epa1361_open/blob/91d2b63942c576aaf617bddedd55286fb3c72a04/final%20assignment/problem_formulation.py#L79-L84

**Comment by quaquel** [comment]
The current implementation is designed to be flexible with respect to the number of planning steps you want to consider. Given that for each planning step you need levers, there is no easy way to specify default values. If you look at the modified notebook, you see my quick trick to work around this. In essence, create a dict with all levers set to 0 and only update those levers you want to change.

**Comment by EwoutH** [comment]
It seems we have a comparable issue in the [Optimization Moro.ipynb](https://github.com/quaquel/epa1361_open/blob/91d2b63942c576aaf617bddedd55286fb3c72a04/final%20assignment/Optimization%20Moro.ipynb) notebook. Could you rerun it with the current version of the EMAworkbench and confirm that it works?

**Comment by EwoutH** [comment]
When I talked to a few other students after the debate today (it was nice btw!), I noticed how they were also unclear about why how the Policy levers were influenced by previous runs in this notebook. I explained shortly that you have to set all the levers explicitly, otherwise they are copied, but maybe this is something worth explaining a bit more extensively in the notebook and mentioning in an announcement of such.
